### PR TITLE
Update SST clipping

### PR DESCRIPTION
**Pull request by marchdf**
**Pull-request type:**
- [ ] Bug fix 
- [ ] Documentation update
- [X] Feature enhancement

## Description of the pull-request

This provides a different SST clipping that fixes spurious TKE solutions near the wall in several SST simulations. This clipping is based on existing clipping strategies in other codes, e.g. CDP. It results in more robust SST simulations.

See attached [pdf](https://github.com/Exawind/nalu-wind/files/5314743/sst-clipping.pdf) for examples of what this fix provides. This fix does not substantially change results in general, as evidenced by the wall hump simulation presented in the pdf as well.

It should be noted that many of the changes provided are only due to `clang-format`. I separated the formatting commit from the actual clipping change for ease of reviewing.

## Checklist

**NOTE** The checklist below is a suggestion and not mandatory. You don't have
to _check all boxes_ to submit a pull request. However, please add a brief
explanation in your pull request summary explaining the omission for the benefit
of reviewers and developers.

*All pull requests*
- [X] Builds successfully (*must test on at least one system/compiler combination*)
  - Operating systems 
    - [X] Linux
    - [X] MacOS
  - Compilers 
    - [ ] GCC
    - [X] LLVM/Clang
    - [X] Intel compilers
    - [ ] NVIDIA CUDA
- [X] Compiles without warnings
- [X] Passes all unit tests
- [X] Passes all regression tests, *except* SST and TAMS regression tests
- [ ] Documentation builds without errors

*For new code updates*
- [ ] Documentation updates - additions to user, theory, and verification manuals
- [ ] New unit tests providing coverage for new code, bug fixes etc.
- [ ] New regression tests exercising the new code

### CODE REVIEW

File: src/ShearStressTransportEquationSystem.C [hunk 1]
```diff
@@ -400,40 +351,41 @@ ShearStressTransportEquationSystem::update_and_clip()
...
-        tkeNp1.get(mi, 0) = viscosity.get(mi, 0) * sdrNew / density.get(mi, 0);
-      } else {
-        // Only SDR is off; reset turbulent viscosity to molecular visc and
-        // compute new SDR based on others
-        turbVisc.get(mi, 0) = viscosity.get(mi, 0);
-        tkeNp1.get(mi, 0) = tkeNew;
-        sdrNp1.get(mi, 0) = density.get(mi, 0) * tkeNew / viscosity.get(mi, 0);
-      }
+      tkeNp1.get(mi, 0) = stk::math::max(tkeNew, tkeMinValue_);
+      sdrNp1.get(mi, 0) = stk::math::max( stk::math::min(sdrNew, sdrMaxValue_), sdrMinValue_);
```
Comment (sayerhs) [inline src/ShearStressTransportEquationSystem.C]: GPU lambda implicitly captures `this`

File: src/ShearStressTransportEquationSystem.C [hunk 2]
```diff
@@ -400,40 +351,41 @@ ShearStressTransportEquationSystem::update_and_clip()
...
+  sdr.sync_to_device();
+
+  nalu_ngp::run_entity_algorithm(
+    "SST::clip", ngpMesh, stk::topology::NODE_RANK, sel,
+    KOKKOS_LAMBDA(const nalu_ngp::NGPMeshTraits<>::MeshIndex& mi) {
+      const double tkeNew = tke.get(mi, 0);
+      const double sdrNew = sdr.get(mi, 0);
+
+      tke.get(mi, 0) = stk::math::max(tkeNew, tkeMinValue_);
+      sdr.get(mi, 0) = stk::math::max( stk::math::min(sdrNew, sdrMaxValue_), sdrMinValue_);
```
Comment (sayerhs) [inline src/ShearStressTransportEquationSystem.C]: GPU lambda implicitly captures `this`

**Review by sayerhs — CHANGES_REQUESTED** [review]

File: src/ShearStressTransportEquationSystem.C [hunk 3]
```diff
@@ -381,59 +335,57 @@ ShearStressTransportEquationSystem::update_and_clip()
...
+    fieldMgr.get_field<double>(sdrEqSys_->wTmp_->mesh_meta_data_ordinal());
 
   auto* turbViscosity = meta.get_field<ScalarFieldType>(
     stk::topology::NODE_RANK, "turbulent_viscosity");
 
   const stk::mesh::Selector sel =
-    (meta.locally_owned_part() | meta.globally_shared_part())
-    & stk::mesh::selectField(*turbViscosity);
+    (meta.locally_owned_part() | meta.globally_shared_part()) &
+    stk::mesh::selectField(*turbViscosity);
```
Comment (sayerhs) [inline src/ShearStressTransportEquationSystem.C:350]:
```suggestion
    (meta.locally_owned_part() | meta.globally_shared_part()) &
    stk::mesh::selectField(*turbViscosity);
    
    // Bring class variables to local scope for lambda capture
    const double tkeMinVal = tkeMinValue_;
    const double sdrMinVal = sdrMinValue_;
    const double sdrMaxVal = sdrMaxValue_;
```

File: src/ShearStressTransportEquationSystem.C [hunk 4]
```diff
@@ -381,59 +335,57 @@ ShearStressTransportEquationSystem::update_and_clip()
...
-      } else {
-        // Only SDR is off; reset turbulent viscosity to molecular visc and
-        // compute new SDR based on others
-        turbVisc.get(mi, 0) = viscosity.get(mi, 0);
-        tkeNp1.get(mi, 0) = tkeNew;
-        sdrNp1.get(mi, 0) = density.get(mi, 0) * tkeNew / viscosity.get(mi, 0);
-      }
+      tkeNp1.get(mi, 0) = stk::math::max(tkeNew, tkeMinValue_);
+      sdrNp1.get(mi, 0) =
+        stk::math::max(stk::math::min(sdrNew, sdrMaxValue_), sdrMinValue_);
```
Comment (sayerhs) [inline src/ShearStressTransportEquationSystem.C:360]:
```suggestion
      tkeNp1.get(mi, 0) = stk::math::max(tkeNew, tkeMinVal);
      sdrNp1.get(mi, 0) =
        stk::math::max(stk::math::min(sdrNew, sdrMaxVal), sdrMinVal);
```

File: src/ShearStressTransportEquationSystem.C [hunk 5]
```diff
@@ -400,40 +351,41 @@ ShearStressTransportEquationSystem::update_and_clip()
...
-        tkeNp1.get(mi, 0) = tkeNew;
-        sdrNp1.get(mi, 0) = density.get(mi, 0) * tkeNew / viscosity.get(mi, 0);
-      }
+      tkeNp1.get(mi, 0) = stk::math::max(tkeNew, tkeMinValue_);
+      sdrNp1.get(mi, 0) = stk::math::max( stk::math::min(sdrNew, sdrMaxValue_), sdrMinValue_);
     });
 
   tkeNp1.modify_on_device();
   sdrNp1.modify_on_device();
-  turbVisc.modify_on_device();
```
Comment (jamelvin) [inline src/ShearStressTransportEquationSystem.C:436]: did we determine why `turbVisc` needed to be updated here?

File: include/ShearStressTransportEquationSystem.h [hunk 6]
```diff
@@ -79,6 +84,10 @@ class ShearStressTransportEquationSystem : public EquationSystem {
   std::vector<stk::mesh::Part *> wallBcPart_;
 
   bool resetTAMSAverages_;     
+
+  const double tkeMinValue_{1.0e-8};
+  const double sdrMinValue_{1.0e-8};
+  const double sdrMaxValue_{1.0e12};
```
Comment (jamelvin) [inline include/ShearStressTransportEquationSystem.h:85]: Sorry, I realize this was in the CDP snippet input file I copied, but it's not actually used in the CDP code anywhere.  I don't know if it really does anything here as I'm not sure SDR gets that big, but we might want to remove it.

File: src/ShearStressTransportEquationSystem.C (hunk 5 above)
Comment (sayerhs) [reply src/ShearStressTransportEquationSystem.C:436]: @jamelvin In the old code turbVisc was reset to laminar values during clipping

Comment (jamelvin) [reply src/ShearStressTransportEquationSystem.C:436]: Yes, sorry I meant, does `turbVisc` get used between here and where it is updated in `LowMach` and thus will it be inconsistent with the clipped values in something, like post-processing?

Comment (marchdf) [reply src/ShearStressTransportEquationSystem.C:436]: I should probably remove turbVisc there... good catch

Comment (sayerhs) [reply src/ShearStressTransportEquationSystem.C:436]: `SurfaceForceAndMomentAlgorithm` uses the effective viscosity not the turbulent viscosity. Even in the previous clipping logic, the effective viscosity didn't reflect the updates to TKE/SDR after the solve.

Comment (jamelvin) [reply src/ShearStressTransportEquationSystem.C:436]: okay, sounds good, thanks!

File: include/ShearStressTransportEquationSystem.h (hunk 6 above)
Comment (marchdf) [reply include/ShearStressTransportEquationSystem.h]: I removed the max clipping.

**Review by jamelvin — APPROVED** [review]

**Pull request merged**